### DISCUSSION TITLE
Error '20240122204029-214'

### DISCUSSION THREAD

**Discussion by meliora-solution** [JonPSmith/AuthPermissions.AspNetCore#104] [General]
Failed to create a new tenant due to an internal error. Contact the support team and provide the string '20240122204029-214' to help them fix your problem.

I am trying to create a multitenant for IdentityUser only.  So I just follow along with an example 2 but then I follow the code of example 3 to create a tenant.  The idea is just create a permission and then create tenant manually because I dont need the invoices tables.  if using example 3, I have no problem creating a tenant or invite user in blazor .  

What is Error :  '20240122204029-214' ??  Could you tell me how to fix this error ?  This error occurred when I tried to create a tenant.

Here is what I register on program.cs

```
builder.Services.RegisterAuthPermissions<UserPermissions>(options =>
{
   
    options.UseLocksToUpdateGlobalResources = false;
    options.TenantType = TenantTypes.SingleLevel;
    options.LinkToTenantType = LinkToTenantTypes.OnlyAppUsers;
    options.EncryptionKey = builder.Configuration[nameof(AuthPermissionsOptions.EncryptionKey)];
    options.PathToFolderToLock = builder.Environment.WebRootPath;

})
 

    .UsingEfCoreSqlServer(connectionString) //NOTE: This uses the same database as the individual accounts DB
    .IndividualAccountsAuthentication()
    .AddSuperUserToIndividualAccounts()
    .RegisterFindUserInfoService<IndividualAccountUserLookup>()
    .AddRolesPermissionsIfEmpty(AppAuthSetupData.RolesDefinition)
    .AddAuthUsersIfEmpty(AppAuthSetupData.UsersRolesDefinition)
    .SetupAspNetCoreAndDatabase(options =>
    {
        //Migrate individual account database
        options.RegisterServiceToRunInJob<StartupServiceMigrateAnyDbContext<IdentityUser100DbContext>>();
        //Add demo users to the database
        options.RegisterServiceToRunInJob<StartupServicesIndividualAccountsAddDemoUsers>();
    });

//manually add services from the AuthPermissions.SupportCode project
//Add the SupportCode services
builder.Services.AddTransient<IAddNewUserManager, IndividualUserAddUserManager<IdentityUser>>();
builder.Services.AddTransient<ISignInAndCreateTenant, SignInAndCreateTenant>();
builder.Services.AddTransient<IInviteNewUserService, InviteNewUserService>();

var app = builder.Build();
```

**Comment by meliora-solution** [comment]
I tried using the autp library download from github 6.1.0, It works.  The error is when I used the library from Nuget Package 6.2.0 

I have to check in detail why the errors occured.  Is it because 6.2.0 or because of Nuget Package ???

**Comment by meliora-solution** [comment]
I change to Nuget Package 6.1.0. It works.  I can create tenant.  So the problem is with Nuget Package 6.2.0.

For now I am just using 6.1.0 .

**Comment by JonPSmith** [comment]
Hi @meliora-solution,

OK, the "Failed to create a new tenant due to an internal error. Contact the support team and provide the string '20240122204029-214' to help them fix your problem." says that something went wrong outside the "Sign up for a new tenant" code. Typically this in your [tenant change service](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Building-a-tenant-change-service). You say you don't want to create a tenant, but you still need to provide a implementation of `ITenantChangeService` service, which doesn't throws exceptions.

[This section](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Sign-up-for-a-new-tenant%2C-with-versioning#what-to-do-if-there-an-fatal-error-during-a-sign-up-a-new-tenant) describes how the '20240122204029-214' helps to find the exception.

You might also want to read the [Why I changed the SignInAndCreateTenant service?](https://github.com/JonPSmith/AuthPermissions.AspNetCore/blob/main/UpdateToVersion620.md#why-i-changed-the-signinandcreatetenant-service) to

**Reply by meliora-solution** [reply]
Thank you John,

I change to Nuget Package 6.1.0. It works now. Ok I read the link you gave. I dont need sharding then I just use Nuget package 6.1.0 and because I have an error using 6.2.0.

What I am trying to do, is to create a multitenant for UserManagement (Identity User), So I dont need the invoices. The idea is to have a single sign on for usermanagement.

So What I do is :
1. follow along the example 2.
2. Then I follow the example 3 to create tenant using code.

It works now. I can create tenant. Now I will try to invite user to join.

If everything went well, I will share it with you on GIthub. 

Here is the code to create a tenant

```
@page "/tenantCreate"

@using AuthPermissions.AspNetCore
@using AuthPermissions.SupportCode.AddUsersServices
@using System.ComponentModel.DataAnnotations

@using AuthpServices.PermissionsCode
@using UserManagement.Components.Account.Shared

<h3>TenantCreate</h3>

@attribute [HasPermission(UserPermissions.TenantCreate)]


@inject ISignInAndCreateTenant userRegisterInvite;

<FluentGrid>
      <FluentGridItem xs="8" sm="4">
      <StatusMessage Message="@Message" />
        <EditForm Model="Input" asp-route-returnUrl="@ReturnUrl" method="post" OnValidSubmit="RegisterTenant" FormName="register">
            <DataAnnotationsValidator />
            <h2>Create a new tenant.</h2>
            <hr />
            <FluentValidationSummary class="text-danger" role="alert" />
             <FluentStack Orientation="Orientation.Vertical">
                <FluentTextField Name="Input.TenantName" @bind-Value="Input.TenantName" AutoComplete="TenantName" Required="true"  Label="Tenant Name" Style="width: 100%" />
                <ValidationMessage For="() => Input.TenantName" class="text-danger" />

                <FluentTextField Name="Input.Email" @bind-Value="Input.Email" AutoComplete="username" Required="true" Placeholder="name@example.com" Label="Email" Style="width: 100%" />
                <ValidationMessage For="() => Input.Email" class="text-danger" />
                <FluentTextField type="password" Name="Input.Password" @bind-Value="Input.Password" AutoComplete="current-password" Required="true" Placeholder="password" Label="Password" Style="width: 100%" />
                <ValidationMessage For="() => Input.Password" class="text-danger" />
                  <FluentButton Type="ButtonType.Submit" Appearance="Appearance.Accent" Style="width: 100%">Create Tenant</FluentButton>
             </FluentStack>
       </EditForm>
      </FluentGridItem>
</FluentGrid>
@code {


    private string? Message = string.Empty;

    [SupplyParameterFromForm]
    private InputModel Input { get; set; } = new();
    [SupplyParameterFromQuery]
    private string? ReturnUrl { get; set; }

    public async Task RegisterTenant(EditContext editContext)
    {
        Input.Version = "Pro";
        var newUserData = new AddNewUserDto { Email = Input.Email, Password = Input.Password, IsPersistent = false};
        var newTenantData = new AddNewTenantDto { TenantName = Input.TenantName, Version = Input.Version};

        var proVersionSetupData = new MultiTenantVersionData
            {
                TenantRolesForEachVersion = new Dictionary<string, List<string>>
                {
                    { "Pro", new List<string> { "Tenant Admin" } }
                },
                TenantAdminRoles = new Dictionary<string, List<string>>
                {
                    { "Pro", new List<string> { "Tenant Admin" } }
                }
            };
        var status = await userRegisterInvite.SignUpNewTenantWithVersionAsync(newUserData, newTenantData,
              proVersionSetupData);

        if (status.HasErrors)
        {
            Message = status.GetAllErrors();
            return;
        }

        Input = new();

        Message = "Tenant Created";

    }

    private sealed class InputModel
    {
        [Required]
        [Display(Name = "tenantName")]
        public string TenantName { get; set; } = "";
        
        [Required]
        [EmailAddress]
        [Display(Name = "Email")]
        public string Email { get; set; } = "";

        [Required]
        [StringLength(100, ErrorMessage = "The {0} must be at least {2} and at max {1} characters long.", MinimumLength = 6)]
        [DataType(DataType.Password)]
        [Display(Name = "Password")]
        public string Password { get; set; } = "";

        public string Version { get; set; } = "Pro";
       
    }
}

```

**Comment by meliora-solution** [comment]
Invite User.  In real application, this should be sent to an Email.

![inviteUser](https://github.com/JonPSmith/AuthPermissions.AspNetCore/assets/150670966/dd738be1-a99f-4af7-a5eb-4892626f1025)

**Reply by JonPSmith** [reply]
There many ways to sent then invite to the user so its up to you to select the best way to send the invite and implement that code.

**Reply by meliora-solution** [reply]
Yes , that is true.  Thank you so much for the autp library.

**Comment by meliora-solution** [comment]
Accept Invitation if pass the verification will create a user of a tenant.  What confused me is that this application is still able to login even user Email is not confirmed.  Because in program.cs the user can only login only is the email is confirmed.  Still have to find way to make EmailConfirmed=true if the verification pass.

```
builder.Services.AddIdentityCore<IdentityUser>(options => options.SignIn.RequireConfirmedAccount = true)
    .AddEntityFrameworkStores<IdentityUser100DbContext>()
    .AddSignInManager()
    .AddDefaultTokenProviders();
```

![acceptInvitation](https://github.com/JonPSmith/AuthPermissions.AspNetCore/assets/150670966/b989133f-8ff2-4827-933a-8444966224a8)

**Comment by meliora-solution** [comment]
Hi @JonPSmith 

I see your examples that when user login and does not have a permission, the page link is not visible. I still dont know how to that in blazor. So when user does not have permission and click the page, it will be redirect to "Account/AccessDenied" page. 

Could you show / point me in what part that you of your code that you do it ?

**Comment by meliora-solution** [comment]
I found the solution how to hide unauthorized page.

```
 @if (user.HasPermission(UserPermissions.TenantList))
 {
   <FluentNavLink Href="tenantlist" Icon="@(new Icons.Regular.Size20.PeopleCommunity())" IconColor="Color.Accent">Tenant List</FluentNavLink>
 }
 @if (user.HasPermission(UserPermissions.TenantCreate))
 {
      <FluentNavLink Href="tenantCreate" Icon="@(new Icons.Regular.Size20.PeopleCommunityAdd())" IconColor="Color.Accent">Tenant Create</FluentNavLink>
      <FluentNavLink Href="tenantCreateSendEmail" Icon="@(new Icons.Regular.Size20.PeopleCommunityAdd())" IconColor="Color.Accent">Tenant Create By Email</FluentNavLink>
 }
 @if (user.HasPermission(UserPermissions.InviteUsers))
 {
     <FluentNavLink Href="inviteUser" Icon="@(new Icons.Regular.Size20.PeopleCommunityAdd())" IconColor="Color.Accent">Invite User</FluentNavLink>
     <FluentNavLink Href="inviteUserByEmail" Icon="@(new Icons.Regular.Size20.PeopleCommunityAdd())" IconColor="Color.Accent">Invite User By Email</FluentNavLink>

 }
```
